File: solutions/haskell/026-longest-valid-parentheses.hs
{-
  Problem 26: Longest Valid Parentheses (LeetCode 32)
  Difficulty: Hard
  Language: Haskell
-}
longestValidParentheses :: String -> Int
longestValidParentheses s = max (scan s '(' ')') (scan (reverse s) ')' '(')
  where
    scan cs open close = go cs 0 0 0
    go [] _ _ mx = mx
    go (c:cs) l r mx
      | c == open  = let l' = l + 1
                     in if l' == r then go cs l' r (max mx (2 * r))
                        else go cs l' r mx
      | otherwise  = let r' = r + 1
                     in if r' > l then go cs 0 0 mx
                        else if l == r' then go cs l r' (max mx (2 * r'))
                        else go cs l r' mx

File: solutions/elixir/026-longest-valid-parentheses.ex
# Problem 26: Longest Valid Parentheses (LeetCode 32)
# Difficulty: Hard
# Language: Elixir
# 
defmodule LongestValidParentheses do
  def longest_valid_parentheses(s) do
    chars = String.graphemes(s)
    max(scan(chars, "(", ")"), scan(Enum.reverse(chars), ")", "("))
  end

  defp scan(chars, open, close) do
    {_, _, mx} =
      Enum.reduce(chars, {0, 0, 0}, fn c, {l, r, mx} ->
        {l, r} = if c == open, do: {l + 1, r}, else: {l, r + 1}
        cond do
          r > l -> {0, 0, mx}
          l == r -> {l, r, max(mx, 2 * r)}
          true -> {l, r, mx}
        end
      end)
    mx
  end
end
